### PR TITLE
Project structure and folder cleanup

### PR DESCRIPTION
**Pull request by MikeFarrington**

### CODE REVIEW

**Comment by MikeFarrington** [comment]
@Ken98045 It looks big, with 138 files changed, but only a dozen files were actually changed, the rest were moved into folders and a few were slightly renamed.

This brings the project in line with a more traditional file/folder layout.  It's not 100% there, but it's a good start.

**Comment by MikeFarrington** [comment]
When you merge this, you'll probably want to do a "squash and merge" to keep the commit history a little cleaner.  I should have done that myself...  I'm still learning how to work on a forked project in GitHub.

**Comment by Ken98045** [comment]
I haven’t used GitHub much because previous employers used other tech.   I definitely should have branched in prep for a merge, but to date nobody else seemed interested in getting involved.   So, if I make dumb GitHub move please let me know.

**Comment by MikeFarrington** [comment]
Git/GitHub definitely has a steep learning curve, that's for sure.  I only started using it about 4-5 years ago myself.  Before that, it was all systems based on a check-in/check-out model.

**Comment by Ken98045** [comment]
The changes I made today were reasonably substantial, and I haven’t even started  thinking about a merge.  I know, very bad practice but I’ve had the luxury of being lazy up to this point.

 

The previous use of binary serialized settings files was again, lazy.  The use of the registry should allow me to add data members without breaking things as badly.  That was a real issue because I hesitated to make needed changes due to breaking other peoples settings every time I needed something else.   There are still a few settings in the Settings.Settings file but I didn’t want to get rid of everything at once.  I do want to get rid of those ASAP.  Maybe I’ll just use the Settings.Settings stuff for defaults or maybe I’ll nuke it completely.   In any event, I want to stop saving there at all before making any new releases.

 

I posted this project originally because I thought I had enough useful personal stuff that I wanted to share it.  It was not originally intended to be a long term project, but then came Covid and I got really bored.

 

BTW, I know that one person had issues installing/running on Windows Server (actually Data Center).  If you have any thoughts on that I’d appreciate it.  I’m 95% sure it is security settings, but I’m not quite sure what is the specific problem.  If you can help me avoid installing a Sever VM it would be appreciated.

**Comment by MikeFarrington** [comment]
I'm running it on Windows Server myself without issue.  

I did run into one issue, and maybe it's the same as theirs...

After cloning your repo and grabbing a database to use, I found myself unable to run because the non-configurable connection string specifies a common default instance name of MSSQLLocalDB.  I also had SQL LocalDB 2016 installed onto my development machine with that instance name.  Thus, SQL 2019 LocalDB could not use it, and SQL 2016 LocalDB (which is what the program was connecting to) wouldn't allow the newer database.

Luckily I knew I had nothing of value that was using SQL 2016 LocalDB, and I went ahead and uninstalled it then changed the SQL 2019 LocalDB instance name to MSSQLLocalDB.  Not everyone will have that luxury as they may be using software that needs that instance name.

That is also one of the reasons I wanted to move the project over to WiX Toolset so that the InstanceName could be specified during installation of SQL 2019 LocalDB.  I imagine we'd want to use an instance name of OnGuard.  

For this reason, the connection string needs to be moved to the registry to allow for modification.  That way if someone else runs into this issue, they can create an different instance for 2019, then point the connection string to it.

Lastly, I know SQL Server throws up serious warnings when you attempt to install onto a Windows Server that is acting as a Domain Controller.  I don't know if that's the case with the LocalDB edition as well.  It could be that the installer failed to install SQL 2019 LocalDB for this reason.  In which case, the end-user would need to download and run that installer manually, and tell the installer that they wish to install despite its objections.

**Comment by MikeFarrington** [comment]
You may not be aware of how git works, but...   when it comes to merging in my code, as long as it says there's no conflicts, then there's nothing manual about merging.  You don't have to make any decisions or anything, just check the code for quality and bring it in.   The "Bird" one is super-simple in that respect.

My bigger PR, this one here, I have "rebased" my changes onto yours.  That means I have already sucked in your changes, handled the merge conflicts.  What the rebase does is makes it look like ALL of my code was written after your code.  It makes for much cleaner repo histories than willy-nilly merging.

So, if you look through the code in this PR, you'll see all of your changes for the registry settings.  Don't look at the "changed files", that just shows my changes.  Because of the aforementioned rebase, all you see is my changes.  However, if you dive down into the branch itself ([MikeFarrington:feature/FolderCleanup](https://github.com/MikeFarrington/On-Guard/tree/feature/FolderCleanup)), not the PR, you'll see that your registry changes are there.

**Comment by MikeFarrington** [comment]
| but then came Covid and I got really bored.

Yup, that's the only reason I've joined your project.  I've never worked on an open-source project before because I've always been too busy with other stuff.  Well, after gearing up with a bunch of cameras and building an NVR, I needed something "fun" to make it better.  I saw your project, thought it had potential, and dove in.

**Comment by MikeFarrington** [comment]
No need to squash, I finally learned how to squash locally in git (I've only ever relied on the squash capabilities in GitHub in repositories I control). So now everything is in one clean commit.

**Comment by Ken98045** [comment]
Thanks!  I’ll need to get serious about learning GitHub.

 

So, I downloaded the Code zip file and I’m getting errors related the missing SQL files on the Build All.  

The rest of the code seems to build fine.  However, when I attempt to open any designer file I get hundreds of errors such as:


Severity

Code

Description

Project

File

Line

Suppression State


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


 

 

 

 

 

 

 

 

 

 

Basically it doesn’t seem to be finding any of the system types.  When I look at references I do see those system files listed.   I have not tried to track these down.  If you are available and can save me some time that would be great!

**Comment by Ken98045** [comment]
OK, the hundreds of errors disappeared on me like magic.  I’m not sure what VS is doing, but whatever it was was annoying

 

From: Ken Koch [mailto:me@aperson.me] 
Sent: Saturday, January 16, 2021 11:46 AM
To: 'Ken98045/On-Guard'
Subject: RE: [Ken98045/On-Guard] Project structure and folder cleanup (#8)

 

Thanks!  I’ll need to get serious about learning GitHub.

 

So, I downloaded the Code zip file and I’m getting errors related the missing SQL files on the Build All.  

The rest of the code seems to build fine.  However, when I attempt to open any designer file I get hundreds of errors such as:


Severity

Code

Description

Project

File

Line

Suppression State


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

31

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

49

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

54

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

59

Active


Error

CS0518

Predefined type 'System.Void' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0518

Predefined type 'System.Object' is not defined or imported

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


Error

CS0246

The type or namespace name 'EventArgs' could not be found (are you missing a using directive or an assembly reference?)

AIDisplay

C:\OnGuard-3\On-Guard-master\AddUrlDlialog.cs

64

Active


 

 

 

 

 

 

 

 

 

 

Basically it doesn’t seem to be finding any of the system types.  When I look at references I do see those system files listed.   I have not tried to track these down.  If you are available and can save me some time that would be great!

**Comment by Ken98045** [comment]
And when I attempt to start debugging I get:



 

After looking at a variety of Google search result I see a lot of “solutions”, but right now nothing is working for me.   It seems like maybe the root cause is that you are using an earlier version of Visual Studio?  What version are you using.  I am at VS Community 2019 build 16.8..4 (which is the latest).

**Comment by Ken98045** [comment]
OK, I have to say my Visual Studio is just plain weird!  I downloaded your changes.  It said it built everything.  Then, later it said there were hundreds/thousands of errors, then it settled down and without changing everything it “just worked/built”.   Then, the next time I tried to run in Debug mode it threw errors related to NewtonSoft, then I built and it maybe just worked.  Then, it stopped working.   So, I’m uninstalling the whole Visual Studio and re-installing.  I don’t trust Repair to work correctly.

 

In the mean time….

**Comment by Ken98045** [comment]
Yes, I’m seeing them there.  I just need to get VS rei-installed and see if that fixes my problems.  I was having breaking performance issue.  For instance, if I tried to create a new WPF project I wouldn’t see a window for about 10 minutes.  NUGet just flat out wouldn’t work for me.  A variety of other things going on that I just couldn’t explain.  After I downloaded your .zip file things just stopped building or built so slowly everything was misleading.  

 

As soon as I’m able to test your changes I will.

**Comment by Ken98045** [comment]
OK, I did resolve my Visual Studio issues by re-install.  However, I do see a problem with your solution/zip file.  The DBMotion files are referenced in the project and in the installer, but they are missing.   I suggest that they be put in a new folder src\database.

The other issue I had was that the newtonsoft.json NuGet reference didn’t work for me.  I had to uninstall it and re-install it.  Then it worked.  I do intend to remove the newtonsoft stuff and use System.Text.Json and the serializers.  That will remove one less complication.  I think I copy/pasted the newtonsoft stuff from some Deepstack sample code.

 

One of the many things that the VS re-install solved was that I can now use NuGet and actually have it work.

 

Also, I’m not seeing the changes you referenced regarding bird detection.  BTW, I do need to create a User Control with a scrolling panel for the object specification stuff or preferably convert that data to a table for them.  It started out very small and grew.

**Comment by Ken98045** [comment]
Sorry for some duplication with a previous message.

 

Please see below

**Comment by Ken98045** [comment]
I’ve been thinking about the Settings.Settings stuff.  While the stuff in there is mostly harmless defaults (with the exception of the connection string), it probably wouldn’t hurt to use them just as defaults as long as they aren’t written to.   While they all could be put in app.config, I don’t think that would solve much.  

So, my current plan is 

1.        Read the initial value (and probably final) value of the connection string from app.config until/unless your installer project can come up with another solution.   The only problem I see there is that I have a minor preference to putting the DB files (if used) someplace other than a VirtualStore directory.

2.       Leave the default values of everything else configurable via UI from the Settings.Settings, but never write to it.  Every setting configured will be written to the registry.

3.       Figure out a better solution for installing/not installing localdb.  I don’t think SQL Express solves anything.  Maybe your setup can just see if there is a usable instance of SQL?

4.       Also, we probably need to give the option to use SQL with an (encrypted) user name/password instead of Windows Authentication for those users who want to do that.  I did think about that but I didn’t want to add a setup dialog entry that 95% of the people would have no idea what to do with it.

 

Let me know what you think.  Because you’re in this early your opinion and input really matters.  Maybe we can smooth this project out but keep the project fun.

**Comment by Ken98045** [comment]
OK, I’ve taken your code base and put them in a new local directory

1.       I fixed the missing DBMotion database files.

2.       Remove all “setting” of Settings.Settings.  Everything other than defaults should now come from the registry.  This makes it much easier to look at for debugging, and in general is probably safe.

3.       Fixed several bugs related to the registry save/load of settings that I introduced.

4.       Fixed a few other bugs.

 

My question for you is: What is the “proper” github method for uploading/merging these files?  I could just commit them and push them up, but I’m not sure if that screws with the history.

 

I know you probably have a life (if one constrained by CoVid.  So, if I don’t hear from you this afternoon (say by 6:00) I’ll just commit them.

**Comment by MikeFarrington** [comment]
Sorry, I wasn't really online this weekend.

The build errors due to the missing DBMotionFiles is because your .gitignore file has exclusions for SQL .MDF/.LDF files on line 250.  It is "bad form" to have large binary files in git.  In order for me to start working on the project, I had to go grab these database files from my installation.  This was one of the reasons I was talking about looking towards EntityFramework...  it'll create the database on project launch.

I am using Visual Studio 16.8.4 (Professional Edition).  Here's a Visual Studio tip for you...  if things go really wonky like they did for you, you want to clear the visual studio cache for your solution.  You need to delete the .VS subfolder in the solution folder (while Visual Studio is closed of course).

You mentioned that you did not see the Bird changes.  That is because those are in a different pull request.  It is good form to have isolated feature branches / pull requests.

**Comment by Ken98045** [comment]
See below.

**Comment by Ken98045** [comment]
FYI – I committed a new version 1.5.2.  The main change was to allow any DeepStack recognized object to be used for analysis.  In addition there are a couple of groups (all vehicles and all mammals).  Several bugs were also fixed as a part of the move of settings to the registry.

 

Let me know if you have a chance to test it.  I did a reasonable amount of runtime testing, but not much Github related testing because it is getting too late for that.

**Comment by Ken98045** [comment]
If you are interested in getting back on this project I could really use help in 2 areas.  

 

The first is rather boring, but I need urgent help getting the installer in shape.

 

The second could prove much more interesting.  I would like to provide the feature of creating areas that are something other than rectangles.  This could be polygons,  it could involve circles too, or even bit, by bit definitions.  You could work on this as a completely separate branch with the code all your own.  It is a 2 part project itelf.  First, a UI for defining areas, and second establishing whether an object overlapps one of these areas.

 

I know it isn’t Kosher for me to “assign” you parts of the project, but at least the second should be interesting, I can easily keep myself out of these areas, and they would offer significant value.

 

I am largely working on code cleanup right now.  This could take a while since the code grew daily without an overall plan and is a real mess.  I will probably offer some low complexity additional features, but should leave the previously described areas alone.  Another area I’ll probably work on part time is moving the code to .NET Core.

 

I do have some stuff ready for a checkin within a few days.

 

If you are no longer interested in working on it, please let me know that too.  I wouldn’t blame you too much because of the code mess.  In addition, this is my first Open Source project, and I’ve probably made all kinds of mistakes with organization, procedures, and courtesies.
